todo-list-backend/tests/integration/graphql-api.spec.ts

```diff
@@ -0,0 +1,526 @@
+
+            const testClient = createTestClient(testServer);
+            query = testClient.query;
+            mutate = testClient.mutate;
+        });
+
+        describe("readTodos", () => {
+            it("raises an error", async () => {
+                let response = await query({
+                    query: READ_TODOS
+                });
+
+                expect(response.errors).toHaveLength(1);
+                expect(response.data).toBeNull();
+            });
+        });
+    });
+});
+
+describe("mutate", () => {
+    const authSecret: string = "This is a perfect secret for testing!";
+
+    describe("given correct user credentials", () => {
+        const storage: IStorage = buildTwoUserInMemoryStorage();
+        const testServer = buildApolloServer(storage, authSecret, false, {});
+        const {query, mutate} = createTestClient(testServer);
+
+        describe("login", () => {
+            it("returns a jwt", async () => {
+                let tokenResponse = await mutate({
+                    mutation: LOGIN,
+                    variables: {
+                        name: "ralph",
+                        password: "ralph"
+                    }
+                });
+
+                expect(tokenResponse.data.login).toBeDefined();
+            });
+        });
+    });
+
+    describe("given wrong user credentials", () => {
+        const storage: IStorage = buildTwoUserInMemoryStorage();
+        const testServer = buildApolloServer(storage, authSecret, false, {});
+        const {query, mutate} = createTestClient(testServer);
+
+        describe("login", () => {
+            it("returns null", async () => {
+                let tokenResponse = await mutate({
+                    mutation: LOGIN,
+                    variables: {
+                        name: "ralph",
+                        password: "false"
+                    }
+                });
+
+                expect(tokenResponse.data.login).toBeNull();
+            });
+        });
+    });
+
+    describe("given a legit authorization token", () => {
+        let user: User;
+        let storage: IStorage;
+        let query: any;
+        let mutate: any;
+
+        beforeEach(async () => {
+            storage = buildTwoUserInMemoryStorage();
+
+            // workaround for the shortcomings of the apollo-server-testing
+            // package as it doesn't generate a real request and therefor a
+            // "real" context will never be generated
+            user = storage.readUser("ralph");
+            const token = sign({username: user.name}, authSecret, {expiresIn: "1 day"});
+            const context = {token: token, user: user};
+
+            const testServer = buildApolloServer(storage, authSecret, false, context);
+
+            const testClient = createTestClient(testServer);
+            query = testClient.query;
+            mutate = testClient.mutate;
```

Comment:
If you have that same block over and over again with slight differences, you could add that at the top of the file:

```js
const setup = (what, ever, option, you, want) => {
}
```

And then in your beforeEachs:
```js
beforeEach(setup)
```
or:
```js
beforeEach(() => {
  dosomethingSpecial = () => { }
  setup(dosomethingSpecial)
})
```